paderbox/io/new_subdir.py

```diff
@@ -0,0 +1,341 @@
+            The id naming that is used for the folder name.
+             - str: 'index':
+                The largest index in basedir + 1.
+                e.g.: '1', '2', ...
+             - str: 'time': A timestamp with the format %Y-%m-%d-%H-%M-%S
+                e.g. '2020-08-13-17-02-57'
+             - callable: Each call should generate a new name.
+        mkdir:
+            Creates the dir and makes the program process/thread safe.
+            Note this option ensures that you don't get a
+            conflict between two concurrent calls of get_new_folder.
+            Example:
+                You launch several times your programs and each should get
+                another folder (e.g. hyperparameter search). When inspecting
+                basedir maybe some recognize they can use '2' as sub folder.
+                This option ensures, that only one program gets the '2' and the
+                remaining programs search for another free id.
+        prefix:
+            Optional prefix for the id. e.g.: '2' -> '{prefix}_2'
+        suffix:
+            Optional suffix for the id. e.g.: '2' -> '2_{suffix}'
+        consider_mpi:
+            If True, only search on one mpi process for the folder and
+            distribute the folder name.
+            When using mpi with this option disables the following can/will
+            happen
+             - When mkdir is True every process will get another folder.
+               i.e. each process has a folder just for this process.
+             - When mkdir is False it is likely that all get the same folder,
+               but this is not guarantied. Never use mpi with consider_mpi and
+               mkdir disabled.
+        dry_run:
+            Similar to mkdir. When true, disable mkdir and print the folder
+            name.
+
+    Returns:
+        pathlib.Path of the new subdir
+
+
+    >>> get_new_subdir('/', dry_run=True)  # root folder usually contain no digits
+    dry_run: "os.mkdir(/1)"
+    PosixPath('/1')
+
+    >>> import numpy as np
+    >>> np.random.seed(0)  # This is for doctest. Never use it in practise.
+    >>> get_new_subdir('/', id_naming=NameGenerator(), dry_run=True)
+    dry_run: "os.mkdir(/helpful_tomato_finch)"
+    PosixPath('/helpful_tomato_finch')
+    """
+
+    if consider_mpi:
+        import dlp_mpi
+        if dlp_mpi.IS_MASTER:
+            pass
+        else:
+            new_folder = None
+            new_folder = dlp_mpi.bcast(new_folder)
+            return new_folder
+
+    basedir = Path(basedir).expanduser().resolve()
+    if not basedir.exists():
+        if dry_run:
+            print(f'dry_run: "os.makedirs({basedir})"')
+            # ToDo: Make this working.
+            #       Will fail when calling os.listdir
+        else:
+            basedir.mkdir(parents=True)
+
+    if Path('/net') in basedir.parents:
+        # If nt filesystem, assert not in /net/home
+        assert Path('/net/home') not in basedir.parents, basedir
```

Comment:
Do we want to have this hard-coded in here or flexible as a parameter for easier use in other environments?

Reply:
I am also not a fan for hard-coded things, but this one in special:
 - We recommend our students to code in the home directory, while storing the experiment files on a different file system.
    - We do this, because the home directory has an expensive backup system that is useful for code (i.e. never lose your written code).
    - Many other frameworks (e.g. Kaldi, ESPnet, Asteroid, ...) have the code and experiment files in the same folder, so I expect this split is special for us.
      - So others often don't care about this split.
- The naming of `/net/home` is very special. I don't expect that anyone will get a problem with this line. 
   - When someone reports, that a name collision, I will erase this assert.